### PR TITLE
krt: fix register and sync race cause double events

### PR DESCRIPTION
**Pull request by howardjohn**
https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/49794/unit-tests_istio/1767212848146026496

honestly the fix is a little gross but it works... with a few other open PRs to fix bad tests, this is looking 100% stable on local runs

### CODE REVIEW

**Comment by ilrudie** [comment]
Thanks for working on this fix. I hit some racy double event issues working through the unit tests on the use-waypoint PR.

File: pkg/kube/krt/collection.go [hunk 1]
```diff
@@ -357,17 +359,27 @@ func newManyCollection[I, O any](cc Collection[I], hf TransformationMulti[I, O],
...
+		h.recomputeMu.Lock()
 		h.eventHandlers.MarkInitialized()
-		handlerReg := c.RegisterBatch(func(events []Event[I]) {
+		handlerReg := c.RegisterBatch(func(events []Event[I], initialSync bool) {
 			if log.DebugEnabled() {
 				h.log.WithLabels("dep", "primary", "batch", len(events)).
 					Debugf("got event")
 			}
-			h.onPrimaryInputEvent(events)
+			// Lock after the initial sync only
```
Comment (bleggett) [inline pkg/kube/krt/collection.go:373]: Yeah, I see what you mean about this being a little weird. Documented tho, and I can't think of a better way to solve it ATM.

**Review by bleggett — APPROVED** [review]

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
Can you fix it in krt collections without influencing the general EventHandler

File: pkg/kube/controllers/common.go [hunk 2]
```diff
@@ -302,13 +302,16 @@ func IgnoreNotFound(err error) error {
 
 // EventHandler mirrors ResourceEventHandlerFuncs, but takes typed T objects instead of any.
 type EventHandler[T Object] struct {
-	AddFunc    func(obj T)
-	UpdateFunc func(oldObj, newObj T)
-	DeleteFunc func(obj T)
+	AddFunc         func(obj T)
+	AddExtendedFunc func(obj T, initialSync bool)
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply pkg/kube/controllers/common.go:306]: without this we will need to copy all of EventHandler into krt which uses this. 

Fwiw the upstream client-go has the same thing for Add (accepts 2 different signatures)

File: pkg/kube/krt/collection.go [hunk 3]
```diff
@@ -473,11 +485,14 @@ func (h *manyCollection[I, O]) Register(f func(o Event[O])) Syncer {
...
 
-func (h *manyCollection[I, O]) RegisterBatch(f func(o []Event[O]), runExistingState bool) Syncer {
-	if !h.eventHandlers.Insert(f) && runExistingState {
-		// Already started. Pause everything, and run through the handler.
+func (h *manyCollection[I, O]) RegisterBatch(f func(o []Event[O], initialSync bool), runExistingState bool) Syncer {
+	if runExistingState {
 		h.recomputeMu.Lock()
 		defer h.recomputeMu.Unlock()
+	}
+	init := h.eventHandlers.Insert(f)
```
Comment (hzxuzhonghu) [inline pkg/kube/krt/collection.go:493]: Isn't this always return false as `MarkInitialized` is called before `RegisterBatch`

In addition, the `init` =  !o.init, should be negative

Comment (howardjohn) [reply pkg/kube/krt/collection.go:493]: > Isn't this always return false as MarkInitialized is called before RegisterBatch



Its not. RegisterBatch can be called from other places. The specific race is for things _outside_ krt calling RegisterBatch and krt itself calling RegisterBatch, which was resulting in double events.

Good call on the naming, will fix

**Review by stevenctl — APPROVED** [review]

File: pkg/kube/controllers/common.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/kube/controllers/common.go:306]: @howardjohn

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/kube/controllers/common.go:306]: IC, it is not only for users, but also for krt inner usage

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pkg/kube/controllers/common.go:306]: So can we document this should not be used by user. Cause i can see for user registered handler,  initialSync is always = false

**Review by hzxuzhonghu — APPROVED** [review]
LGTM

**Pull request merged**